### ISSUE TITLE
TabPFNClassifier fails when dataset contains pd.NA values

### ISSUE THREAD

**Issue by VIVEK-MARRI** [PriorLabs/TabPFN#545]
### Describe the bug

While using TabPFNClassifier with categorical columns containing pd.NA, the model fails or behaves inconsistently.
It seems that internal preprocessing doesn’t handle missing values properly in some cases, especially with categorical data.
This issue appears even without explicit preprocessing by the user.

### Steps/Code to Reproduce

import pandas as pd
from tabpfn import TabPFNClassifier
from sklearn.model_selection import train_test_split

# Minimal dataset with missing categorical value
data = {
    "feature1": ["a", "b", pd.NA, "d"],
    "feature2": [1, 2, 3, 4],
    "target": [0, 1, 0, 1],
}
df = pd.DataFrame(data)

X = df[["feature1", "feature2"]]
y = df["target"]

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)

clf = TabPFNClassifier(device="cpu")

# Model should handle NA internally
clf.fit(X_train, y_train)
preds = clf.predict(X_test)
print(preds)

### Expected Results

The classifier should handle pd.NA values internally (numeric or categorical) and produce valid predictions without requiring manual preprocessing (e.g., fillna or encoding).
No error should be thrown.

### Actual Results

When fitting or predicting on data containing pd.NA, TabPFNClassifier either throws an error or produces inconsistent predictions.
The issue seems to originate from internal preprocessing that doesn’t handle missing values correctly.

### Versions

```shell
OS: Windows 11
Python: 3.13
TabPFN: Latest (installed from GitHub main branch)
Device: CPU
```

**Comment by oscarkey** [comment]
Hi Vivek, thanks for the reproduction script!

For me it seems to work well, but maybe I am missing something? I am also using MacOS rather than Windows.

I use the script:
```python
import pandas as pd
from tabpfn import TabPFNClassifier
from sklearn.model_selection import train_test_split

data = {
    "feature1": ["a", "b", pd.NA, "d"],
    "feature2": [1, 2, 3, 4],
    "target": [0, 1, 0, 1],
}
df = pd.DataFrame(data)

X = df[["feature1", "feature2"]]
y = df["target"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42
)

print("X_train\n", X_train)
print("y_train\n", y_train)
print("X_test\n", X_test)
print("y_test\n", y_test)
clf = TabPFNClassifier(device="cpu")
clf.fit(X_train, y_train)
preds = clf.predict(X_test)
print("preds:", preds)
```
Which gives the output:
```
$ uv run -p 3.13 --with https://github.com/PriorLabs/TabPFN.git python bug_repro.py
X_train
   feature1  feature2
3        d         4
0        a         1
2     <NA>         3
y_train
 3    1
0    0
2    0
Name: target, dtype: int64
X_test
   feature1  feature2
1        b         2
y_test
 1    1
Name: target, dtype: int64
preds: [0]
```

I also tried setting `random_state=46`, which meant the NA was in `X_test`, and it still seemed to work:
```
$ uv run -p 3.13 --with https://github.com/PriorLabs/TabPFN.git python bug_repro.py
X_train
   feature1  feature2
0        a         1
3        d         4
1        b         2
y_train
 0    0
3    1
1    1
Name: target, dtype: int64
X_test
   feature1  feature2
2     <NA>         3
y_test
 2    0
Name: target, dtype: int64
preds: [1]
```

Could you post the error message you are seeing? Or explain a bit more what you mean by "inconsistent predictions"?

**Comment by VIVEK-MARRI** [comment]
Thanks for testing this on macOS and running the reproduction script.  
I also ran it on **Windows 11** with **Python 3.13**, and here are my results:

---

### Case 1 — `random_state = 42` (NA in train set)
```python
X_train
   feature1  feature2
3        d         4
0        a         1
2     <NA>         3
y_train
 3    1
0    0
2    0
Name: target, dtype: int64

X_test
   feature1  feature2
1        b         2
y_test
 1    1
Name: target, dtype: int64

preds: [0]
```

 ---

### Case 2 — `random_state = 46` (NA in test set)

```python
X_train
   feature1  feature2
0        a         1
3        d         4
1        b         2
y_train
 0    0
3    1
1    1
Name: target, dtype: int64

X_test
   feature1  feature2
2     <NA>         3
y_test
 2    0
Name: target, dtype: int64

preds: [1]
```

### Observation

Both cases run without errors on Windows.

However, when `pd.NA` appears in the test set (Case 2), the prediction differs (`[1]`) compared to when NA is in train (Case 1, `[0]`).

This suggests that handling of missing values might be slightly nondeterministic or platform/environment-dependent.

Would it help if I also share the versions of dependencies (pandas, numpy) or rerun with more random states to check consistency?

Happy to provide more details.

**Comment by oscarkey** [comment]
Maybe I am missing something, but isn't this the behaviour we expect? There are very few data points in this example, so I think it's expected that the prediction will change when the split changes, and this would happen even without the na values.

**Comment by VIVEK-MARRI** [comment]
Thanks for the clarification!  

Yes, I understand that with such a small dataset, predictions can naturally vary between splits. My main goal was to verify that the model runs without errors when `pd.NA` values are present in both train and test sets.  

I just wanted to highlight that the behavior is consistent in terms of not raising errors, even if predictions differ due to the small dataset size.

**Comment by oscarkey** [comment]
okay great! Sounds like everything is working then? I'll close the issue, but feel free to reopen if something is still broken :)
